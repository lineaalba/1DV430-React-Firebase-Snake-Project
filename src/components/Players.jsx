/**
 * @author Filippa Jakobsson
 * @version 1.0
 */

import React, { useState, useEffect } from 'react'
import firebase from '../firebase'

/**
* Gets relevant data from firestore, order by highest score.
* @returns {array}
*/
const OnlinePlayers = () => {
    const [players, setPlayers] = useState([])
    useEffect(() => { 
        firebase
            .firestore()
            .collection('players')
            .orderBy('highscore', 'desc')
            .onSnapshot((snapshot) => {
                const newPlayer = snapshot.docs.map((doc) => ({
                    id: doc.id,
                    ...doc.data()
                }))

                setPlayers(newPlayer)
            })
    }, [])
    return players
}

/**
* Presents online player name, score and game area.
*/
const Players = () => {
    const players = OnlinePlayers()

        if (players.length > 0) {
            return (
                <div>
                    <ol>
                        {players.map(player => (
                            <li className="players__list" key={player.id}>
                                <h3>{player.name} | Score: {player.highscore}</h3>
                                <div className="game__area" style={{background: `${player.background}`}}>
                                    <div className="snake__food" style={{left: `${player.food[0]}%`, top: `${player.food[1]}%`}}></div>
                                    <div>
                                        {
                                            JSON.parse(player.snake).map((dot, i) => {
                                                const styles = {
                                                        left: `${dot[0]}%`,
                                                        top: `${dot[1]}%`,
                                                        background: `${player.snakeColor}`
                                                }
                                                return (  
                                                    <div className="snake__dot" key={i} style={styles}></div>        
                                                )
                                    
                                               
                                        })}
                                        
                                    </div>
                                </div>
                            </li>
                        ))}
                    </ol>
                </div>
            )
        } else {
            return (
                <div>
                <br></br>
                    <h4>There are no players online right now.</h4>
                <br></br>
                </div>
            )
        }
    }

export default Players
