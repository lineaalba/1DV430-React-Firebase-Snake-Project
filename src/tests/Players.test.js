import React from 'react'
import { render } from '@testing-library/react'
import { shallow } from 'enzyme'
import Players from '../components/Players'

describe('Tests the players component', () => {

  it('matches snapshot', () => {
    const wrapper = shallow(<Players />)
    expect(wrapper.html()).toMatchSnapshot()
  })
 
  it('renders a text if no one is playing', () => {
    const { getByText } = render(<Players/>)
    expect(getByText('There are no players online right now.')).toBeInTheDocument()
  })
})

