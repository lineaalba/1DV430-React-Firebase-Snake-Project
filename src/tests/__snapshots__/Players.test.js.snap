// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests the players component matches snapshot 1`] = `"<div><br/><h4>There are no players online right now.</h4><br/></div>"`;
